src/generator.cpp

```diff
@@ -16,1743 +16,280 @@ limitations under the License.
 
-GeneratorEquationPtr Generator::GeneratorImpl::processNode(const XmlNodePtr &node,
-                                                           const ComponentPtr &component)
+/**
+ * @brief The Generator::GeneratorImpl struct.
+ *
+ * The private implementation for the Generator class.
+ */
+struct Generator::GeneratorImpl
 {
-    // Create and keep track of the equation associated with the given node.
-
-    GeneratorEquationPtr equation = std::make_shared<GeneratorEquation>(component);
-
-    mEquations.push_back(equation);
-
-    // Actually process the node and return its corresponding equation.
+    Generator *mGenerator = nullptr;
 
-    processNode(node, equation->mAst, equation->mAst->mParent.lock(), component, equation);
+    AnalyserModelWeakPtr mModel;
+    AnalyserModelPtr mLockedModel;
 
-    return equation;
-}
+    std::string mCode;
 
-void Generator::GeneratorImpl::processComponent(const ComponentPtr &component)
-{
-    // Retrieve the math string associated with the given component and process
-    // it, one equation at a time.
+    GeneratorProfilePtr mOwnedProfile = libcellml::GeneratorProfile::create();
+    GeneratorProfileWeakPtr mProfile;
+    GeneratorProfilePtr mLockedProfile;
 
-    XmlDocPtr xmlDoc = std::make_shared<XmlDoc>();
-    std::string math = component->math();
+    bool retrieveLockedModelAndProfile();
+    void resetLockedModelAndProfile();
 
-    if (!math.empty()) {
-        xmlDoc->parseMathML(math, false);
+    AnalyserVariablePtr analyserVariable(const VariablePtr &variable) const;
 
-        XmlNodePtr mathNode = xmlDoc->rootNode();
+    double scalingFactor(const VariablePtr &variable) const;
 
-        for (XmlNodePtr node = mathNode->firstChild(); node != nullptr; node = node->next()) {
-            if (node->isMathmlElement()) {
-                processNode(node, component);
-            }
-        }
-    }
+    bool isRelationalOperator(const AnalyserEquationAstPtr &ast) const;
+    bool isAndOperator(const AnalyserEquationAstPtr &ast) const;
+    bool isOrOperator(const AnalyserEquationAstPtr &ast) const;
+    bool isXorOperator(const AnalyserEquationAstPtr &ast) const;
+    bool isLogicalOperator(const AnalyserEquationAstPtr &ast) const;
+    bool isPlusOperator(const AnalyserEquationAstPtr &ast) const;
+    bool isMinusOperator(const AnalyserEquationAstPtr &ast) const;
+    bool isTimesOperator(const AnalyserEquationAstPtr &ast) const;
+    bool isDivideOperator(const AnalyserEquationAstPtr &ast) const;
+    bool isPowerOperator(const AnalyserEquationAstPtr &ast) const;
+    bool isRootOperator(const AnalyserEquationAstPtr &ast) const;
+    bool isPiecewiseStatement(const AnalyserEquationAstPtr &ast) const;
 
-    // Go through the given component's variables and make sure that everything
-    // makes sense.
+    std::string replace(std::string string, const std::string &from,
```

Comment:
This might be useful to other places? Maybe it should be in the utilities file(s)?

Reply:
Agreed. Made that method one of our utilities in commit 489939a3.